pkg/controller/queuejob/metrics.go

```diff
@@ -0,0 +1,45 @@
+/*
+Copyright 2023 The Multi-Cluster App Dispatcher Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queuejob
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+func registerMetrics(controller *XController) {
+	allocatableCapacityCpu := prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+		Subsystem: "mcad",
+		Name:      "allocatable_capacity_cpu",
+		Help:      "Allocatable CPU Capacity (in milicores)",
+	}, func() float64 { return controller.GetAllocatableCapacity().MilliCPU })
+	metrics.Registry.MustRegister(allocatableCapacityCpu)
+
+	allocatableCapacityMemory := prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+		Subsystem: "mcad",
+		Name:      "allocatable_capacity_memory",
+		Help:      "Allocatable Memory Capacity",
+	}, func() float64 { return controller.GetAllocatableCapacity().Memory })
+	metrics.Registry.MustRegister(allocatableCapacityMemory)
+
+	allocatableCapacityGpu := prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+		Subsystem: "mcad",
+		Name:      "allocatable_capacity_gpu",
+		Help:      "Allocatable GPU Capacity",
+	}, func() float64 { return float64(controller.GetAllocatableCapacity().GPU) })
```

Comment:
The calls to `GetAllocatableCapacity()` are quite expensive to be executed in GaugeFuncs. Should I call it periodically and cache the values like in the original PR?
https://github.com/project-codeflare/multi-cluster-app-dispatcher/pull/573/files#diff-cd034f141bebdabd3ed7e6118e9581ec4646707beaf8438f164ae51f382f2c1dR69-R77

Reply:
I believe caching the values like in the original PR is the right call.

Reply:
Right, listing all the Nodes and all the Pods clearly does not scale at all.

We need to reconsider re-activating / improving the cache, intersecting #588. That may provide a smooth transition once we migrate over to controller-runtime.

Otherwise, we can consider a simple caching mechanism, e.g. that would rate limit the call to `allocatableCapacity` using something like `golang.org/x/time/rate.NewLimiter`.

Reply:
I made another commit to schedule the expensive call to run once a minute. I wasn't sure how to use `golang.org/x/time/rate.NewLimiter` without making it look like an overkill.